NetKet/Supervised/vmc.hpp

```diff
@@ -0,0 +1,424 @@
+// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NETKET_SUPERVISEDVARIATIONALMONTECARLO_HPP
```

Comment:
Related to my general comment on the Overlap, I think we should make an effort to have a single function working with generic loss functions. In that logic, the sampling from vmc (thus this whole class) is not needed.